### PR TITLE
changing sort.grad to work with other ndim

### PR DESCRIPTION
**Pull request by SinaHonari**
fixes issue #3024

### CODE REVIEW

**Comment by nouiz** [comment]
the pr need a rebase.

**Comment by nouiz** [comment]
travis still have this error:
```
/home/travis/miniconda/envs/pyenv/lib/python2.6/site-packages/theano/tensor/sort.py:68:52: W503 line break before binary operator
```

File: theano/tensor/sort.py [hunk 1]
```diff
@@ -62,39 +62,31 @@ def grad(self, inputs, output_grads):
             self, 0, axis,
             "Currently, we only implement the gradient on sort for vector"
             " matrix (and axis is None or 0) and tensor3")
```
Comment (lamblin) [inline theano/tensor/sort.py]: Now, I think you implemented all supported cases. So I would get rid of that default "not implemented" message, and maybe put it in an "else" block below, changing the message to say that this is not a supported case.

File: theano/tensor/sort.py [hunk 2]
```diff
@@ -62,39 +62,31 @@ def grad(self, inputs, output_grads):
...
-                    (isinstance(axis, theano.Constant) and axis.data is None)):
-                rev_idx = self.__get_argsort_indices(a, axis)
-                inp_grad = output_grads[0][rev_idx].reshape(a.shape)
+        if ((isinstance(axis, theano.Constant) or (isinstance(axis,
+                                                   theano.tensor.
+                                                   TensorVariable) and
+                                                   axis.ndim == 0)) and
+                axis.data is not None):
+            indices = self.__get_argsort_indices(a, axis)
+            inp_grad = output_grads[0][tuple(indices)]
```
Comment (lamblin) [inline theano/tensor/sort.py]: Do you think you could call `tuple(...)` before returning the value from `__get_argsort_indices` when axis is not None? That way, these two lines could be combined with the `elif` block.

**Comment by lamblin** [comment]
Apart from that, which may need a bit of refactoring, that looks great, thanks a lot!
So we can keep the symbolic axis, since everything is working with it.

**Comment by lamblin** [comment]
Can you fix the pep8 issue? Then I think it is ready to merge.

**Comment by SinaHonari** [comment]
how to check that specific test locally before committing PR?

**Comment by nouiz** [comment]
nosetests theano/tests/test_flake8.py

But it don't cover exactly the same as in travis, not sure why...

**Comment by lamblin** [comment]
That failure was in the master at the time, so I'll merge.

**Pull request merged**